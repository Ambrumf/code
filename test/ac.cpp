
#include<iostream>
#include<cstdio>
#include<cstring>
#include<algorithm>
#include<sstream>
#include<vector>
#include<map>
#include<queue>
#include<stack>
#include<set>
#include<cmath>
#define up(i, x, y) for(int i = x; i <= y; i++)
#define down(i, x, y) for(int i = x; i >= y; i--)
#define MAXN ((int)1e6 + 10)
#define INF 0x3f3f3f3f
using namespace std;
typedef unsigned long long ull;
int arr[MAXN];
int pow_mod(ull a, ull n, int m) //快速幂取模
{
    ull base = a, ans = 1;
    while(n)
    {
        if(n & 1){
            ans = ans * base % m;
        }
        base = base * base % m;
        n >>= 1;
    }
    return (int)(ans % m);
}
int solve(ull a, ull b, int n)
{
    int len = 1;
    for(int i = 2; 1 ; i++)
    {
        arr[i] = (arr[i - 1] + arr[i - 2]) % n;
        if(arr[i - 1] == arr[0] && arr[i] == arr[1]) {
            len = i - 1; //周期
            break;
        }
    }
    int loc = pow_mod(a % len, b , len); //位置
    return arr[loc];
}
int main()
{
    //freopen("data.in","r",stdin);
    //freopen("ac.out","w",stdout);
    int T; cin>>T;
    while(T--){
        ull a, b;
        int n;
        cin>>a>>b>>n;
        arr[0] = 0, arr[1] = 1 % n;
        cout<<solve(a, b, n)<<'\n';
    }
    return 0;
}