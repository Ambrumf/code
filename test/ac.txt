No enemy spy
No enemy spy
No enemy spy
cmm vky
No enemy spy
No enemy spy
ygy
No enemy spy
No enemy spy
No enemy spy
No enemy spy
No enemy spy
No enemy spy
No enemy spy
No enemy spy
No enemy spy
No enemy spy
No enemy spy
No enemy spy
No enemy spy
uzg
No enemy spy
No enemy spy
No enemy spy
No enemy spy
uox
No enemy spy
No enemy spy
No enemy spy
daq sqi
No enemy spy
No enemy spy
hhr
No enemy spy
No enemy spy
uoa
gcg wwu uyd eae
No enemy spy
No enemy spy
No enemy spy
No enemy spy
No enemy spy
wyd pjh pjp
No enemy spy
vpl spq
txm
No enemy spy
No enemy spy
nns kcd cce
sej ydl noq muy gxg
gnt
No enemy spy
qvh
No enemy spy
No enemy spy
No enemy spy
No enemy spy
kzt
xdi ymk pst ywa
nec xpk tnz
No enemy spy
No enemy spy
fkz
No enemy spy
No enemy spy
No enemy spy
xge
byx
kar fta oho
No enemy spy
myq
No enemy spy
pfp
No enemy spy
jfy
No enemy spy
No enemy spy
No enemy spy
kff
No enemy spy
No enemy spy
No enemy spy
No enemy spy
nme
No enemy spy
No enemy spy
crp
No enemy spy
No enemy spy
rpi
No enemy spy
No enemy spy
No enemy spy
No enemy spy
tqy
No enemy spy
No enemy spy
No enemy spy
No enemy spy
eut
No enemy spy
kxk
mts
No enemy spy
No enemy spy
wat
No enemy spy
lpn
oaj
No enemy spy
umx stv kqp
No enemy spy
No enemy spy
No enemy spy
No enemy spy
No enemy spy
No enemy spy
No enemy spy
vtm
No enemy spy
kcx rjv
No enemy spy
No enemy spy
No enemy spy
No enemy spy
No enemy spy
No enemy spy
oqb
No enemy spy
No enemy spy
yjr sws yrt
No enemy spy
fcm lko
No enemy spy
dvw zzc
No enemy spy
llj max
ujp
No enemy spy
eak
fvc pel
No enemy spy
No enemy spy
No enemy spy
jto
zvx
No enemy spy
zff zcn
No enemy spy
No enemy spy
xmg
No enemy spy
jlo
ygv
jpl lcr
aqi ucp
lxt mwe
xkx gvo gyl
uni
qjq
No enemy spy
xnh
No enemy spy
rnj ggw
syl mhj
No enemy spy
bht zpn
lit
No enemy spy
qfz zux vag
kmr shw cep
No enemy spy
bqv
No enemy spy
rkd
No enemy spy
No enemy spy
qel
No enemy spy
No enemy spy
No enemy spy
No enemy spy
No enemy spy
xeu
yon
aan
bdd jpz
No enemy spy
jxa
rvc
No enemy spy
No enemy spy
No enemy spy
qgy csz upr
No enemy spy
fag oaa nwz
lwi
No enemy spy
No enemy spy
vsm
No enemy spy
kax
No enemy spy
ynt vek
No enemy spy
No enemy spy
cgp
zvh
No enemy spy
No enemy spy
No enemy spy
No enemy spy
No enemy spy
No enemy spy
wbj
No enemy spy
No enemy spy
No enemy spy
No enemy spy
xyt mds uoj
lnw
qfj vim
No enemy spy
stp
No enemy spy
No enemy spy
hlb
dfq nxf kma
No enemy spy
No enemy spy
No enemy spy
No enemy spy
qme
aft
No enemy spy
No enemy spy
wfe
ywj
No enemy spy
No enemy spy
hec szg fkv
No enemy spy
No enemy spy
owd
No enemy spy
No enemy spy
No enemy spy
No enemy spy
No enemy spy
yrp qkm
yij snz
No enemy spy
