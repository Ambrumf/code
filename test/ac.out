1
63
0
70
124
0
140
57
126
864
27
330
684
312
19
77
401
107
5
21
45
182
407
885
595
0
53
342
338
3
411
0
511
0
258
201
21
367
47
753
153
18
567
201
217
59
35
418
0
273
126
57
217
291
3
147
66
220
498
487
117
184
552
1
805
212
512
544
309
55
0
392
115
0
0
189
136
210
81
67
303
330
0
117
0
70
603
181
0
0
0
722
69
299
297
285
233
698
459
0
47
34
135
357
112
165
648
567
32
322
312
8
21
204
444
564
259
137
490
375
55
27
24
0
273
23
475
148
63
484
91
157
239
269
55
0
132
0
51
295
0
239
167
228
57
0
35
37
